### PR TITLE
MMP-232 Implement agent deployment

### PR DESCRIPTION
**Pull request by WilliamPinho**

### CODE REVIEW

File: mule-deployer/src/main/java/org/mule/tools/client/standalone/installer/MuleStandaloneInstaller.java [hunk 1]
```diff
@@ -129,10 +126,7 @@ protected File getDependency(ArtifactDescription artifactDescription)
       artifactResolver.resolve(artifact, mavenProject.getRemoteArtifactRepositories(), localRepository);
       return artifact.getFile();
     } catch (AbstractArtifactResolutionException e) {
-      throw new MojoExecutionException("Couldn't download artifact: " + e.getMessage(), e);
-    } catch (Exception e) {
-      throw new MojoFailureException("Couldn't download artifact: " + e.getMessage());
+      throw new DeploymentException("Couldn't download artifact: " + e.getMessage(), e);
```
Comment (damiansima) [inline mule-deployer/src/main/java/org/mule/tools/client/standalone/installer/MuleStandaloneInstaller.java]: +      throw new DeploymentException("Couldn't download artifact: " + e.getMessage(), e);
vs 
+      throw new DeploymentException("Fail to download artifact: " + artifactDesdription.getName(), e);

Not a good idea to throw an exception with e.getMessage and then send also e as parameter

File: mule-deployer/src/main/java/org/mule/tools/client/AbstractDeployer.java [hunk 2]
```diff
@@ -10,31 +10,51 @@
...
-  private final Log log;
+  protected final DeploymentConfiguration deploymentConfiguration;
+  protected final Log log;
 
-  public AbstractDeployer(String applicationName, File applicationFile, Log log) {
-    this.applicationName = applicationName;
-    this.applicationFile = applicationFile;
+
+
+  public AbstractDeployer(DeploymentConfiguration deploymentConfiguration, Log log) throws DeploymentException {
```
[the remark this answers is not in the document]
Comment (WilliamPinho) [reply mule-deployer/src/main/java/org/mule/tools/client/AbstractDeployer.java]: Perfect! Change to the DeployerLog interface and removed this class completely from the project.

File: mule-deployer/src/main/java/org/mule/tools/client/AbstractDeployerFactory.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+import org.mule.tools.client.agent.AgentDeployer;
+import org.mule.tools.client.arm.ArmDeployer;
+import org.mule.tools.client.cloudhub.CloudhubDeployer;
+import org.mule.tools.client.standalone.deployment.ClusterDeployer;
+import org.mule.tools.client.standalone.deployment.StandaloneDeployer;
+import org.mule.tools.client.standalone.exception.DeploymentException;
+import org.mule.tools.model.DeploymentConfiguration;
+
+
+public class AbstractDeployerFactory {
```
[the remark this answers is not in the document]
Comment (WilliamPinho) [reply mule-deployer/src/main/java/org/mule/tools/client/AbstractDeployerFactory.java]: Done!

File: mule-deployer/src/main/java/org/mule/tools/client/AbstractDeployerFactory.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+  public AbstractDeployer getDeployer(DeploymentConfiguration deploymentConfiguration, Log log)
```
[the remark this answers is not in the document]
Comment (WilliamPinho) [reply mule-deployer/src/main/java/org/mule/tools/client/AbstractDeployerFactory.java]: Done!

File: mule-deployer/src/main/java/org/mule/tools/client/agent/AgentClient.java [hunk 5]
```diff
@@ -35,16 +35,15 @@ public void deployApplication(String applicationName, File file) {
     Response response =
         put(uri, APPLICATIONS_PATH + applicationName, Entity.entity(file, MediaType.APPLICATION_OCTET_STREAM_TYPE));
 
-    if (response.getStatus() != 202) // Created
-    {
+    if (response.getStatus() != ACCEPTED) {
```
[the remark this answers is not in the document]
Comment (WilliamPinho) [reply mule-deployer/src/main/java/org/mule/tools/client/agent/AgentClient.java]: Cheers!

File: mule-deployer/src/main/java/org/mule/tools/client/standalone/deployment/StandaloneDeployer.java [hunk 6]
```diff
@@ -157,4 +135,100 @@ public void addDomainFromDeploymentConfiguration(DeploymentConfiguration configu
...
+    if (!FilenameUtils.getBaseName(deploymentConfiguration.getApplication().getName())
+        .equals(deploymentConfiguration.getApplicationName())) {
+      try {
+        File destApplication =
+            new File(deploymentConfiguration.getApplication().getParentFile(),
+                     deploymentConfiguration.getApplicationName() + ".jar");
+        FileUtils.copyFile(deploymentConfiguration.getApplication(), destApplication);
+        deploymentConfiguration.setApplication(destApplication);
+      } catch (IOException e) {
+        throw new DeploymentException("Couldn't rename [" + deploymentConfiguration.getApplication() + "] to ["
```
[the remark this answers is not in the document]
Comment (WilliamPinho) [reply mule-deployer/src/main/java/org/mule/tools/client/standalone/deployment/StandaloneDeployer.java]: Done!

File: mule-deployer/src/main/java/org/mule/tools/client/standalone/installer/MuleStandaloneInstaller.java (hunk 1 above)
Comment (WilliamPinho) [reply mule-deployer/src/main/java/org/mule/tools/client/standalone/installer/MuleStandaloneInstaller.java]: Done! Thanks for the hint.

**Pull request merged**